### PR TITLE
Add onRamps support to LinkMon

### PR DESCRIPTION
**Pull request by fbac**
- Add support for CCIP onRamps.
- Add support for CL Automation to trigger the contract.

### CODE REVIEW

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 1]
```diff
@@ -95,7 +106,7 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
     address[] calldata addresses,
     uint96[] calldata minBalances,
     uint96[] calldata topUpAmounts
-  ) external onlyOwner {
+  ) external onlyRoleOrAdminRole(EXECUTOR_ROLE) {
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:109]: nitpick, but I think this name is a bit confusing. Can we just call it `onlyAdminOrExecutor`?

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 2]
```diff
@@ -63,24 +63,35 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
     uint16 maxCheck,
     uint8 upkeepInterval
-  ) ConfirmedOwner(msg.sender) {
+  ) {
     if (linkTokenAddress == address(0)) revert InvalidLinkTokenAddress(linkTokenAddress);
-    LINK_TOKEN = IERC20(linkTokenAddress);
+    _setRoleAdmin(ADMIN_ROLE, ADMIN_ROLE);
+    _setRoleAdmin(EXECUTOR_ROLE, ADMIN_ROLE);
+    _grantRole(ADMIN_ROLE, admin);
+    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:93]: iiuc, we are overriding the default admin role for `ADMIN_ROLE` and `EXECUTOR_ROLE`, meaning the the `DEFAULT_ADMIN_ROLE` is now useless and has no privileges. Therefore, I think we can probably avoid setting any users to that role.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 3]
```diff
@@ -63,24 +63,35 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+
   uint256 private s_minWaitPeriodSeconds;
   uint16 private s_maxPerform;
   uint16 private s_maxCheck;
   uint8 private s_upkeepInterval;
   address[] private s_watchList;
   mapping(address targetAddress => MonitoredAddress targetProperties) internal s_targets;
+  mapping(uint64 dstChainSelector => address onRamp) internal s_onRampAddresses;
 
   /// @param linkTokenAddress the LINK token address
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:80]: can we add the rest of the natspec?

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 4]
```diff
@@ -118,6 +129,53 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+    s_targets[targetAddress].isActive = false;
+    for (uint256 i; i < s_watchList.length; i++) {
+      if (s_watchList[i] == targetAddress) {
+        s_watchList[i] = s_watchList[s_watchList.length - 1];
+        s_watchList.pop();
+        return true;
+      }
+    }
+    return false;
+  }
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:177]: this is going to be rather expensive... do we know how large we expect `s_watchlist` to be? If it's going to be larger than, say 10... we should probably find another solution.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 5]
```diff
@@ -158,15 +216,15 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
 
   function topUp(address[] memory targetAddresses) public whenNotPaused {
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:217]: can we add natspec here

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 4 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:177]: I don't think we can have a right estimate on the watchlist length, but this function will be called very sporadically (if at all) and it's here only to ensure CL Automation adds onRamps correctly.

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 6]
```diff
@@ -63,24 +69,50 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+  IERC20 private immutable i_linkToken;
+
   uint256 private s_minWaitPeriodSeconds;
   uint16 private s_maxPerform;
   uint16 private s_maxCheck;
   uint8 private s_upkeepInterval;
-  address[] private s_watchList;
+
+  /// @notice s_watchList contains all the addresses watched by this monitor
+  /// It mainly provides the length() function
```
[the remark this answers is not in the document]
Comment (RyanRHall) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:84]: sorry, poor communication on my part. I meant like this!

```solidity
  /// @notice s_watchList contains all the addresses watched by this monitor
  /// @dev It mainly provides the length() function
```

I think in general, it's nice to give all comments an appropriate natspec heading

**Review by RyanRHall — CHANGES_REQUESTED** [review]

**Review by RyanRHall — APPROVED** [review]

**Pull request merged**